activation/poetdb.go

```diff
@@ -0,0 +1,129 @@
+package activation
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	xdr "github.com/nullstyle/go-xdr/xdr3"
+	"github.com/spacemeshos/go-spacemesh/common"
+	"github.com/spacemeshos/go-spacemesh/database"
+	"github.com/spacemeshos/go-spacemesh/types"
+	"github.com/spacemeshos/merkle-tree"
+	"github.com/spacemeshos/poet/hash"
+	"github.com/spacemeshos/poet/shared"
+	"github.com/spacemeshos/poet/verifier"
+	"github.com/spacemeshos/sha256-simd"
+)
+
+type PoetDb struct {
+	store database.DB
+}
+
+func NewPoetDb(store database.DB) *PoetDb {
+	return &PoetDb{store: store}
+}
+
+func (db *PoetDb) AddMembershipProof(proof types.PoetMembershipProof) error {
+	root, err := calcRoot(proof.Members)
+	if err != nil {
+		return fmt.Errorf("failed to calculate membership root: %v", err)
+	}
+
+	var members bytes.Buffer
+	if _, err := xdr.Marshal(&members, proof.Members); err != nil {
+		return fmt.Errorf("failed to marshal members: %v", err)
+	}
+
+	if err := db.store.Put(root, members.Bytes()); err != nil {
+		return fmt.Errorf("failed to store membership proof: %v", err)
+	}
+
+	return nil
+}
+
+func (db *PoetDb) AddPoetProof(proof types.PoetProof) error {
+	if _, err := db.store.Get(proof.MembershipRoot); err != nil {
+		return fmt.Errorf("failed to fetch matching membership proof: %v", err)
+	}
+	if err := validatePoet(proof.MembershipRoot, proof.MerkleProof, proof.LeafCount); err != nil {
+		return fmt.Errorf("failed to validate poet proof: %v", err)
+	}
+
+	var poetProof bytes.Buffer
+	if _, err := xdr.Marshal(&poetProof, proof); err != nil {
+		return fmt.Errorf("failed to marshal poet proof: %v", err)
+	}
+
+	if err := db.store.Put(proof.Root, poetProof.Bytes()); err != nil {
+		return fmt.Errorf("failed to store poet proof: %v", err)
+	}
+
+	if err := db.store.Put(makeKey(proof.PoetId, proof.RoundId), proof.Root); err != nil {
```

Comment:
I've given this some thought and I don't believe it's an issue. If a malicious poet signs another message with a different proof nothing bad really happens.

This key is only used for miners to detect that the poet proof they're waiting for is ready. Once they find the proof they only reference the ref=hash of the proof and this is the only key that a validator would use to validate the ATX.

If a collision happens in the ref, it's a different story. Originally it would have been hard to influence the ref since it was the root of the merkel proof. Now that it's the hash of the entire message, we may have accidentally made grinding easier.

At the moment, one could theoretically grind this by generating random poet ids or rounds until the resulting message collides with an existing ref. Not sure how we should handle this though. Maybe exclude these fields from the hash?

Reply:
I've now separated the non-proof fields so the second issue is also resolved.